File: _PROJECTS/AURENIX AGENCY/aurenix-monorepo/legacy/verify_security.ts
import { webcrypto } from 'node:crypto';
import { pbkdf2 } from 'node:crypto';
import { promisify } from 'node:util';

const pbkdf2Async = promisify(pbkdf2);

/**
 * Derives the AES-256 key using PBKDF2-SHA256
 */
async function getKey(): Promise<any> {
  const masterKey = process.env.AURENIX_MASTER_KEY;
  if (!masterKey) {
    throw new Error("CRITICAL: AURENIX_MASTER_KEY is missing.");
  }
  
  const saltEnv = process.env.AURENIX_ENCRYPTION_SALT;
  if (!saltEnv) {
    throw new Error("CRITICAL: AURENIX_ENCRYPTION_SALT is missing.");
  }

  const salt = new TextEncoder().encode(saltEnv);
  
  const derivedBits = await pbkdf2Async(
    masterKey,
    salt,
    100000,
    32,
    'sha256'
  );

  return await webcrypto.subtle.importKey(
    'raw',
    derivedBits,
    { name: 'AES-GCM' },
    false,
    ['encrypt', 'decrypt']
  );
}

async function encryptSecret(plainText: string): Promise<string> {
  if (!plainText) return "";

  try {
    const key = await getKey();
    const nonce = webcrypto.getRandomValues(new Uint8Array(12));
    const encodedText = new TextEncoder().encode(plainText);

    const ciphertextBuffer = await webcrypto.subtle.encrypt(
      {
        name: 'AES-GCM',
        iv: nonce,
      },
      key,
      encodedText
    );

    const ciphertextArray = new Uint8Array(ciphertextBuffer);
    const combined = new Uint8Array(nonce.length + ciphertextArray.length);
    combined.set(nonce);
    combined.set(ciphertextArray, nonce.length);

    return Buffer.from(combined).toString('base64url');
  } catch (error) {
    console.error("Encryption failed:", error);
    throw error;
  }
}

async function decryptSecret(encryptedText: string): Promise<string> {
  if (!encryptedText) return "";

  try {
    const key = await getKey();
    
    // base64url decode
    const combined = Buffer.from(encryptedText, 'base64url');
    
    if (combined.length < 12) return "[INVALID_DATA]";

    const nonce = combined.subarray(0, 12);
    const ciphertextWithTag = combined.subarray(12);

    const decryptedBuffer = await webcrypto.subtle.decrypt(
      {
        name: 'AES-GCM',
        iv: nonce,
      },
      key,
      ciphertextWithTag
    );

    return new TextDecoder().decode(decryptedBuffer);
  } catch (error) {
    console.error("Decryption failed:", error);
    return "[DECRYPTION_FAILED]";
  }
}

const masterKey = "test-master-key-1234567890abcdef";
const salt = "test-salt-123";

process.env.AURENIX_MASTER_KEY = masterKey;
process.env.AURENIX_ENCRYPTION_SALT = salt;

async function test() {
  const secret = "SuperSecretValue2026";
  console.log(`Original: ${secret}`);

  try {
    const encrypted = await encryptSecret(secret);
    console.log(`Encrypted (TS): ${encrypted}`);

    const decrypted = await decryptSecret(encrypted);
    console.log(`Decrypted (TS): ${decrypted}`);

    if (secret !== decrypted) {
      console.error("FAIL: TS roundtrip failed");
      process.exit(1);
    }
    console.log("SUCCESS: TS roundtrip passed");

    // Handle CLI args
    const args = process.argv.slice(2);
    if (args.length > 0) {
        const input = args[0];
        if (input.length > 20 && !input.includes(" ")) {
            console.log(`Decrypting input: ${input}`);
            console.log(`Decrypted Input: ${await decryptSecret(input)}`);
        } else {
            console.log(`Encrypting input: ${input}`);
            console.log(`Encrypted Input: ${await encryptSecret(input)}`);
        }
    }

  } catch (e) {
    console.error(e);
    process.exit(1);
  }
}

test();

File: _PROJECTS/AURENIX AGENCY/aurenix-monorepo/legacy/verify_security.py
import sys
import os

# Add agent-worker/src to path for imports
BASE_DIR = os.path.dirname(os.path.abspath(__file__))
SRC_PATH = os.path.join(BASE_DIR, 'apps', 'agent-worker', 'src')
sys.path.append(SRC_PATH)

try:
    from tools.security import encrypt_secret, decrypt_secret
except ImportError as e:
    print(f"Import error: {e}")
    # Fallback to direct path append for tools
    sys.path.append(os.path.join(SRC_PATH, 'tools'))
    from security import encrypt_secret, decrypt_secret

def test_encryption():
    # Set env vars for test
    os.environ["AURENIX_MASTER_KEY"] = "test-master-key-1234567890abcdef"
    os.environ["AURENIX_ENCRYPTION_SALT"] = "test-salt-123"

    secret = "SuperSecretValue2026"
    print(f"Testing basic encryption for: {secret}")

    encrypted = encrypt_secret(secret)
    print(f"Encrypted (Python): {encrypted}")

    decrypted = decrypt_secret(encrypted)
    print(f"Decrypted (Python): {decrypted}")

    if secret != decrypted:
        print("FAIL: Python roundtrip failed")
        sys.exit(1)
    
    print("SUCCESS: Python roundtrip passed")

def test_multitenant_isolation():
    """
    Simulates isolation check. In Aurenix, while we use a Master Key, 
    we must ensure that changing the key or salt correctly prevents decryption.
    True multitenancy in the DB is handled via organization_id filters in Prisma.
    """
    print("\n--- Testing Security Isolation ---")
    os.environ["AURENIX_MASTER_KEY"] = "TENANT_A_KEY_1234567890123456"
    os.environ["AURENIX_ENCRYPTION_SALT"] = "SALT_A"
    
    secret = "DataForTenantA"
    encrypted_a = encrypt_secret(secret)
    
    print("Switching security context to Tenant B...")
    os.environ["AURENIX_MASTER_KEY"] = "TENANT_B_KEY_1234567890123456"
    os.environ["AURENIX_ENCRYPTION_SALT"] = "SALT_B"
    
    decrypted_attempt = decrypt_secret(encrypted_a)
    print(f"Decryption attempt with wrong key: {decrypted_attempt}")
    
    if decrypted_attempt == secret:
        print("CRITICAL FAIL: Isolation breach! Data decrypted with wrong key.")
        sys.exit(1)
    else:
        print("SUCCESS: Isolation confirmed. Wrong key cannot decrypt data.")

if __name__ == "__main__":
    test_encryption()
    test_multitenant_isolation()
    
    # Check args for cross-verification
    if len(sys.argv) > 1:
        # CLI tool mode 
        start_val = sys.argv[1]
        if len(start_val) > 20 and " " not in start_val:
            print(f"Decrypting input: {start_val}")
            print(f"Decrypted Input: {decrypt_secret(start_val)}")
        else:
            print(f"Encrypting input: {start_val}")
            print(f"Encrypted Input: {encrypt_secret(start_val)}")
